packages/editor-ui/src/composables/useContextMenu.ts

```diff
@@ -0,0 +1,240 @@
+import { useDataSchema } from './useDataSchema';
+
+export type ContextMenuTarget =
+	| { source: 'canvas' }
+	| { source: 'node-right-click'; node: INode }
+	| { source: 'node-button'; node: INode };
+export type ContextMenuAction =
+	| 'open'
+	| 'copy'
+	| 'toggle_activation'
+	| 'duplicate'
+	| 'execute'
+	| 'rename'
+	| 'toggle_pin'
+	| 'delete'
+	| 'select_all'
+	| 'deselect_all'
+	| 'add_node'
+	| 'add_sticky';
+
+const position = ref<XYPosition>([0, 0]);
+const isOpen = ref(false);
+const target = ref<ContextMenuTarget>({ source: 'canvas' });
+const actions = ref<IActionDropdownItem[]>([]);
+
+export const useContextMenu = () => {
+	const uiStore = useUIStore();
+	const nodeTypesStore = useNodeTypesStore();
+	const workflowsStore = useWorkflowsStore();
+	const sourceControlStore = useSourceControlStore();
+	const { getInputDataWithPinned } = useDataSchema();
+
+	const i18n = useI18n();
+	const isReadOnly = computed(
+		() => sourceControlStore.preferences.branchReadOnly || uiStore.isReadOnlyView,
+	);
+
+	const canAddNodeOfType = (nodeType: INodeTypeDescription) => {
+		const sameTypeNodes = workflowsStore.allNodes.filter((n) => n.type === nodeType.name);
+		return nodeType.maxNodes === undefined || sameTypeNodes.length < nodeType.maxNodes;
+	};
+
+	const canDuplicateNode = (node: INode): boolean => {
+		const nodeType = nodeTypesStore.getNodeType(node.type, node.typeVersion);
+		if (!nodeType) return false;
+		if (NOT_DUPLICATABE_NODE_TYPES.includes(nodeType.name)) return false;
+
+		return canAddNodeOfType(nodeType);
+	};
+
+	const canPinNode = (node: INode): boolean => {
+		const nodeType = nodeTypesStore.getNodeType(node.type, node.typeVersion);
+		const dataToPin = getInputDataWithPinned(node);
+		if (!nodeType || dataToPin.length === 0) return false;
+		return nodeType.outputs.length === 1 && !PIN_DATA_NODE_TYPES_DENYLIST.includes(node.type);
+	};
+
+	const hasPinData = (node: INode): boolean => {
+		return !!workflowsStore.pinDataByNodeName(node.name);
+	};
+	const close = () => {
+		target.value = { source: 'canvas' };
+		isOpen.value = false;
+		actions.value = [];
+		position.value = [0, 0];
+	};
+
+	const targetNodes = computed(() => {
+		if (!isOpen.value) return [];
+		const selectedNodes = uiStore.selectedNodes.map((node) =>
+			workflowsStore.getNodeByName(node.name),
+		) as INode[];
+		const currentTarget = target.value;
+		if (currentTarget.source === 'canvas') {
+			return selectedNodes;
+		} else if (currentTarget.source === 'node-right-click') {
+			const isNodeInSelection = selectedNodes.some((node) => node.name === currentTarget.node.name);
+			return isNodeInSelection ? selectedNodes : [currentTarget.node];
+		}
+
+		return [currentTarget.node];
+	});
+
+	watch(
+		() => uiStore.nodeViewOffsetPosition,
+		() => {
+			close();
+		},
+	);
+
+	return {
+		isOpen,
+		position,
+		target,
+		actions,
+		targetNodes,
+		open: (event: MouseEvent, menuTarget: ContextMenuTarget = { source: 'canvas' }) => {
```

Comment:
Nitpick: why inlined this only function? It's always better to have them as variables (more future-proof) so you can easily use them when/if they need to call each other.